src/lib/server/lib.c

```diff
@@ -0,0 +1,229 @@
+/*
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/*
+ * $Id$
+ *
+ * @file lib.c
+ * @brief Handle global configuration, initialisation and freeing for libraries
+ *
+ * @copyright 2022 The FreeRADIUS server project
+ */
+#include <freeradius-devel/server/lib.h>
+#include <freeradius-devel/server/main_config.h>
+#include <freeradius-devel/util/atexit.h>
+
+/*
+ *  Terminator for array of fr_lib_autoinit_t
+ */
+fr_lib_autoinit_t const fr_lib_autoinit_terminator = { .name = NULL };
+
+/*
+ *  Global list of libraries
+ */
+static fr_lib_list_t *lib_list;
+
+/** Parse the global config section for a library and call its init function
+ *
+ * @param[in] lib	to configure and initialise
+ * @return
+ *	- 0 on success
+ *	- -1 on failure
+ */
+static int lib_init_call(fr_lib_t *lib)
+{
+	CONF_SECTION	*global_cs, *cs;
+
+	/*
+	 *	Find relevant config section
+	 */
+	global_cs = cf_section_find(cf_root(main_config->root_cs), "global", NULL);
+	if (!global_cs) {
+		ERROR("Missing global config section");
+		return -1;
+	}
+
+	cs = cf_section_find(global_cs, lib->name, NULL);
+	if (!cs) {
+		ERROR("Missing \"%s\" subsection in global config", lib->name);
+		return -1;
+	}
+
+	if ((cf_section_rules_push(cs, lib->autoinit->config)) < 0 ||
+	    (cf_section_parse(lib, lib->autoinit->inst, cs) < 0)) {
+		cf_log_err(cs, "Failed evaluating configuration for libldap");
+		return -1;
+	}
+
+	/*
+	 *  Call the init callback if defined
+	 */
+	if (lib->autoinit->init && (lib->autoinit->init()) < 0) return -1;
+
+	lib->initialised = true;
+
+	return 0;
+}
+
+/** Initialise a list of libraries
+ *
+ * @param to_init	Array of autoinit structures detailing libraries to initialise
+ * @return
+ *	- 0 on succcess
+ *	- -1 on failure
+ */
+static int lib_autoinit(fr_lib_autoinit_t * const *to_init)
+{
+	fr_lib_autoinit_t * const *p = to_init;
+
+	for (; (*p)->name; p++) {
+		fr_lib_t	*lib = NULL;
+
+		lib = fr_rb_find(&lib_list->libs, &(fr_lib_t){ .name = (*p)->name });
+
+		/*
+		 *  If the library is already initialised, just increase the reference count
+		 */
+		if ((lib) && (lib->initialised)) {
```

Comment:
what about `if lib && !lib->initialized` ?  Maybe make that an assert?

Reply:
That's a valid thing which could happen.  When libraries are loaded during config parsing, the entry would be created in the tree on the first occasion the library is loaded, however initialisation can't happen until the config parsing is completed as `main_config->root_cs` is not populated until that point.
So if the same library is used by more than one module in FR which causes it to be loaded more than once during config parsing, after the first time it is loaded, subsequent calls will have `lib && !lib-initialised`.  The initialisation will happen after config parsing has been completed by a call to `lib_init()`

Reply:
Agree with @ndptech.  Multiple modules use things like `libfreeradius-curl`, and this is the global config, i.e. there's one config structure for the entire process.